api-error-response_test.go

```diff
@@ -0,0 +1,279 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage (C) 2015 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required bZy applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package minio
+
+import (
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+	"strconv"
+	"testing"
+)
+
+// Tests validate the Error generator function for http response with error.
+func TestHttpRespToErrorResponse(t *testing.T) {
+	// 'formulateAPIErrorResponse' formulates ErrorResponse for given APIError.
+	// provides a encodable populated response values.
+	formulateAPIErrorResponse := func(err APIError, bucketName string) ErrorResponse {
+		var data = ErrorResponse{}
+		data.Code = err.Code
+		data.Message = err.Description
+
+		data.BucketName = bucketName
+		// TODO implement this in future
+		return data
+	}
+
+	// Encodes the response headers into XML format.
+	encodeErr := func(response interface{}) []byte {
+		var bytesBuffer bytes.Buffer
+		bytesBuffer.WriteString(xml.Header)
+		encode := xml.NewEncoder(&bytesBuffer)
+		encode.Encode(response)
+		return bytesBuffer.Bytes()
+	}
+
+	// `formulteAPIErrorResponse` Mocks XML error response from the server.
+	formulteAPIErrorResponse := func(APIErr APIError, bucketName string) *http.Response {
+		// generate error response.
+		// response body contains the XML error message.
+		resp := &http.Response{}
+		errorResponse := formulateAPIErrorResponse(APIErr, bucketName)
+		encodedErrorResponse := encodeErr(errorResponse)
+		// write Header.
+		resp.StatusCode = APIErr.HTTPStatusCode
+		resp.Body = ioutil.NopCloser(bytes.NewBuffer(encodedErrorResponse))
+
+		return resp
+	}
+
+	// 'formulateErrResponse' contructs error response based http Status Code
+	formulateErrResponse := func(resp *http.Response, code, message, bucketName, objectName string) ErrorResponse {
+		errResp := ErrorResponse{
+			Code:       code,
+			Message:    message,
+			BucketName: bucketName,
+			Key:        objectName,
+			RequestID:  resp.Header.Get("x-amz-request-id"),
+			HostID:     resp.Header.Get("x-amz-id-2"),
+			Region:     resp.Header.Get("x-amz-bucket-region"),
+		}
+		return errResp
+	}
+
+	// Generate invalid argument error.
+	formualteInvalidError := func(message string) error {
```

Comment:
It'll be good if tests are as self contained as possible. This will enable detection bugs due to changes in the source. Yes, there is a separate functional tests for `ErrInvalidArgument` too, but this too is not a bad way to test is what I believe, it'll act as one additional layer of test.